OpenSim/Moco/MocoGoal/MocoImpulseTrackingGoal.cpp

```diff
@@ -0,0 +1,362 @@
+        return 6;
+    }
+
+    // Check the group's alternative frames.
+    // Check each alternative frame path in the order provided. As soon as one
+    // of these paths matches the name of the sphere base frame or
+    // half space base frame, use the contact forces applied to that base frame.
+    const auto& sphereBasePath = sphereBase.getAbsolutePathString();
+    const auto& halfSpaceBasePath = halfSpaceBase.getAbsolutePathString();
+    for (int ia = 0; ia < group.getProperty_alternative_frame_paths().size();
+        ++ia) {
+        const auto& path = group.get_alternative_frame_paths(ia);
+        if (path == sphereBasePath) { return 0; }
+        if (path == halfSpaceBasePath) { return 6; }
+    }
+
+    OPENSIM_THROW_FRMOBJ(Exception,
+        "Contact force '{}' has sphere base frame '{}' and half space base "
+        "frame '{}'. One of these frames should match the applied_to_body "
+        "setting ('{}') of ExternalForce '{}', or match one of the "
+        "alternative_frame_paths, but no match found.",
+        contactForce.getAbsolutePathString(), sphereBaseName,
+        halfSpaceBaseName, appliedToBody, group.get_external_force_name());
+}
+
+void MocoImpulseTrackingGoal::calcIntegrandImpl(
+    const IntegrandInput& input, double& integrand) const {
+    const auto& state = input.state;
+    const auto& time = state.getTime();
+    getModel().realizeVelocity(state);
+    SimTK::Vector timeVec(1, time);
+
+    integrand = 0;
+    SimTK::Vec3 force_ref;
+    for (int ig = 0; ig < (int)m_groups.size(); ++ig) {
+
+        // Get contact groups.
+        const auto& group = m_groups[ig];
+
+        // Model force.
+        SimTK::Vec3 force_model(0);
+        for (const auto& entry : group.contacts) {
+            Array<double> recordValues = entry.first->getRecordValues(state);
+            const auto& recordOffset = entry.second;
+            for (int im = 0; im < force_model.size(); ++im) {
+                force_model[im] += recordValues[recordOffset + im];
+            }
+        }
+
+        // Reference force.
+        for (int ir = 0; ir < force_ref.size(); ++ir) {
+            force_ref[ir] = group.refSplines[ir].calcValue(timeVec);
+        }
+
+        // Re-express the reference force.
+        if (group.refExpressedInFrame) {
+            group.refExpressedInFrame->expressVectorInGround(state, force_ref);
+        }
+
+        // Apply scale factors for this marker, if they exist.
+        const auto& scaleFactorRef = m_scaleFactorRefs[ig];
+        for (int j = 0; j < 3; ++j) {
+            if (scaleFactorRef[j] != nullptr) {
+                force_ref[j] *= scaleFactorRef[j]->getScaleFactor();
+            }
+        }
+
+        SimTK::Vec3 error3D = force_model - force_ref;
+
+        double error = error3D[get_impulse_axis()];
+
+        integrand += error;
+    }
+
+    // Square the integrated impulse error
+    integrand = (integrand * integrand);
```

Comment:
We can't square the integrand here because that would violate the math for force impulse. See the comment in `calcGoalImpl` for where to square.

Reply:
Is this because calcGoalImpl gets called for each mesh point? again...just out of curiosity!

Reply:
`calcGoalImpl` is computed based on the entire trajectory at each iteration. `calcIntegrandImpl` is computed at each mesh point. In `calcGoalImpl`, you have access to the trajectory's initial state, final state, and the integral of whatever you put in `calcIntegrandImpl`. 

Impulse is the integral of force, not force squared, so squaring here would produce some other quantity that isn't impulse.